### ISSUE TITLE
Materializing datasets with metadata files fails

### ISSUE THREAD

**Issue by mvdbeek** [galaxyproject/galaxy#14177]
Try uploading a deferred bam dataset, then run a tool on the deferrred dataset.
That results in
```
Traceback (most recent call last):
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/jobs/runners/__init__.py", line 257, in prepare_job
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/jobs/__init__.py", line 1231, in prepare
    compute_environment = compute_environment or self.default_compute_environment(job)
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/tools/evaluation.py", line 141, in set_compute_environment
    materialized_objects = self._materialize_objects(deferred_objects, self.local_working_directory)
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/tools/evaluation.py", line 234, in _materialize_objects
    undeferred = dataset_materializer.ensure_materialized(value)
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/model/deferred.py", line 165, in ensure_materialized
    materialized_dataset_instance.set_meta()
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/model/__init__.py", line 4079, in set_meta
    return self.datatype.set_meta(self, **kwd)
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/datatypes/binary.py", line 741, in set_meta
    if index_flag == "-b":
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/model/__init__.py", line 8354, in file_name
    identifier = getattr(self, store_by)
TypeError: getattr(): attribute name must be string
```
(at least when celery is not enabled, will try that next)

**Comment by mvdbeek** [comment]
This is the sentry link with more context: https://sentry.galaxyproject.org/organizations/galaxy/issues/85952/?project=3&query=is%3Aunresolved&statsPeriod=14d

**Comment by mvdbeek** [comment]
Fixed in #14181
